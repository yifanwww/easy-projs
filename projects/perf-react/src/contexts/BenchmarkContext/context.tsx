import { useConst } from '@easy-pkg/hooks';
import { abstractFn } from '@easy-pkg/utils';
import { useImmerReducer } from '@easy-pkg/utils-react';
import type { ImmerReducer } from '@easy-pkg/utils-react';
import type { Draft } from 'immer';
import { createContext } from 'react';

import type { BenchmarkResult, BenchmarkTypes, ComponentName } from 'src/common/benchmark';

import { benchmarkResultAdapter, benchmarkResultSelector } from './adapters';
import type { GroupBenchmarkResults, IBenchmarkContext, IBenchmarkContextUpdaters } from './types';

const initialContext: IBenchmarkContext = {
    mount: {
        average: { noHooks: 0, useCallback: 0, useMemo: 0, useReducer: 0, useRef: 0, useState: 0 },
        noHooks: benchmarkResultAdapter.getInitialState(),
        useCallback: benchmarkResultAdapter.getInitialState(),
        useMemo: benchmarkResultAdapter.getInitialState(),
        useReducer: benchmarkResultAdapter.getInitialState(),
        useRef: benchmarkResultAdapter.getInitialState(),
        useState: benchmarkResultAdapter.getInitialState(),
    },
    unmount: {
        average: { noHooks: 0, useCallback: 0, useMemo: 0, useReducer: 0, useRef: 0, useState: 0 },
        noHooks: benchmarkResultAdapter.getInitialState(),
        useCallback: benchmarkResultAdapter.getInitialState(),
        useMemo: benchmarkResultAdapter.getInitialState(),
        useReducer: benchmarkResultAdapter.getInitialState(),
        useRef: benchmarkResultAdapter.getInitialState(),
        useState: benchmarkResultAdapter.getInitialState(),
    },
    update: {
        average: { noHooks: 0, useCallback: 0, useMemo: 0, useReducer: 0, useRef: 0, useState: 0 },
        noHooks: benchmarkResultAdapter.getInitialState(),
        useCallback: benchmarkResultAdapter.getInitialState(),
        useMemo: benchmarkResultAdapter.getInitialState(),
        useReducer: benchmarkResultAdapter.getInitialState(),
        useRef: benchmarkResultAdapter.getInitialState(),
        useState: benchmarkResultAdapter.getInitialState(),
    },
    totalResults: benchmarkResultAdapter.getInitialState(),
};

function calculateAverageStats(state: Draft<IBenchmarkContext>): void {
    function _calculate(group: GroupBenchmarkResults, name: ComponentName): void {
        const total = benchmarkResultSelector.selectAll(group[name]).reduce((prev, curr) => prev + curr.stats.mean, 0);
        group.average[name] = total / benchmarkResultSelector.selectTotal(group[name]);
    }

    _calculate(state.mount, 'noHooks');
    _calculate(state.mount, 'useCallback');
    _calculate(state.mount, 'useMemo');
    _calculate(state.mount, 'useReducer');
    _calculate(state.mount, 'useRef');
    _calculate(state.mount, 'useState');

    _calculate(state.unmount, 'noHooks');
    _calculate(state.unmount, 'useCallback');
    _calculate(state.unmount, 'useMemo');
    _calculate(state.unmount, 'useReducer');
    _calculate(state.unmount, 'useRef');
    _calculate(state.unmount, 'useState');

    _calculate(state.update, 'noHooks');
    _calculate(state.update, 'useCallback');
    _calculate(state.update, 'useMemo');
    _calculate(state.update, 'useReducer');
    _calculate(state.update, 'useRef');
    _calculate(state.update, 'useState');
}

type IBenchmarkAction =
    | { type: 'add'; result: BenchmarkResult }
    | { type: 'clear'; benchmarkType: BenchmarkTypes; componentName: ComponentName }
    | { type: 'clear-all' };

const reducer: ImmerReducer<IBenchmarkContext, IBenchmarkAction> = (state, action) => {
    let never: never;
    switch (action.type) {
        case 'add': {
            const { result } = action;

            benchmarkResultAdapter.addOne(state.totalResults, result);
            benchmarkResultAdapter.addOne(state[result.type][result.name], result);

            break;
        }

        case 'clear': {
            const { benchmarkType, componentName } = action;

            const selectIds = benchmarkResultSelector.selectIds(state.totalResults).filter((id) => {
                const result = benchmarkResultSelector.selectById(state.totalResults, id)!;
                return result.name === componentName && result.type === benchmarkType;
            });

            benchmarkResultAdapter.removeMany(state.totalResults, selectIds);
            benchmarkResultAdapter.removeAll(state[benchmarkType][componentName]);

            break;
        }

        case 'clear-all':
            benchmarkResultAdapter.removeAll(state.totalResults);
            for (const name in state.mount) benchmarkResultAdapter.removeAll(state.mount[name as ComponentName]);
            for (const name in state.unmount) benchmarkResultAdapter.removeAll(state.unmount[name as ComponentName]);
            for (const name in state.update) benchmarkResultAdapter.removeAll(state.update[name as ComponentName]);
            break;

        default:
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            never = action;
    }

    calculateAverageStats(state);
};

export const BenchmarkContext = createContext<IBenchmarkContext>(initialContext);

export const BenchmarkContextUpdater = createContext<IBenchmarkContextUpdaters>({
    add: abstractFn,
    clear: abstractFn,
    clearAll: abstractFn,
});

export const BenchmarkProvider: React.FC = ({ children }) => {
    const [context, dispatch] = useImmerReducer(reducer, initialContext);

    const updaters = useConst<IBenchmarkContextUpdaters>(() => ({
        add: (result) => dispatch({ type: 'add', result }),
        clear: (benchmarkType, componentName) => dispatch({ type: 'clear', benchmarkType, componentName }),
        clearAll: () => dispatch({ type: 'clear-all' }),
    }));

    return (
        <BenchmarkContextUpdater.Provider value={updaters}>
            <BenchmarkContext.Provider value={context}>{children}</BenchmarkContext.Provider>
        </BenchmarkContextUpdater.Provider>
    );
};
