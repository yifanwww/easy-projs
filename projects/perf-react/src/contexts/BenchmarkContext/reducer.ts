import type { Draft } from 'immer';
import type { ImmerReducer } from 'use-immer';

import type { BenchmarkResult, BenchmarkTypes, ComponentName } from 'src/common/benchmark';

import { benchmarkResultAdapter, benchmarkResultSelector } from './adapters';
import type { BenchmarkContextState, GroupBenchmarkResults } from './types';

function calculateAverageStats(state: Draft<BenchmarkContextState>): void {
    function _calculate(group: GroupBenchmarkResults, name: ComponentName): void {
        const total = benchmarkResultSelector.selectAll(group[name]).reduce((prev, curr) => prev + curr.stats.mean, 0);
        group.average[name] = total / benchmarkResultSelector.selectTotal(group[name]);
    }

    _calculate(state.mount, 'noHooks');
    _calculate(state.mount, 'useCallback');
    _calculate(state.mount, 'useMemo');
    _calculate(state.mount, 'useReducer');
    _calculate(state.mount, 'useRef');
    _calculate(state.mount, 'useState');

    _calculate(state.unmount, 'noHooks');
    _calculate(state.unmount, 'useCallback');
    _calculate(state.unmount, 'useMemo');
    _calculate(state.unmount, 'useReducer');
    _calculate(state.unmount, 'useRef');
    _calculate(state.unmount, 'useState');

    _calculate(state.update, 'noHooks');
    _calculate(state.update, 'useCallback');
    _calculate(state.update, 'useMemo');
    _calculate(state.update, 'useReducer');
    _calculate(state.update, 'useRef');
    _calculate(state.update, 'useState');
}

type ReducerAction =
    | { type: 'add'; result: BenchmarkResult }
    | { type: 'clear'; benchmarkType: BenchmarkTypes; componentName: ComponentName }
    | { type: 'clear-all' };

export const reducer: ImmerReducer<BenchmarkContextState, ReducerAction> = (state, action) => {
    let never: never;
    switch (action.type) {
        case 'add': {
            const { result } = action;

            benchmarkResultAdapter.addOne(state.totalResults, result);
            benchmarkResultAdapter.addOne(state[result.type][result.name], result);

            break;
        }

        case 'clear': {
            const { benchmarkType, componentName } = action;

            const selectIds = benchmarkResultSelector.selectIds(state.totalResults).filter((id) => {
                const result = benchmarkResultSelector.selectById(state.totalResults, id)!;
                return result.name === componentName && result.type === benchmarkType;
            });

            benchmarkResultAdapter.removeMany(state.totalResults, selectIds);
            benchmarkResultAdapter.removeAll(state[benchmarkType][componentName]);

            break;
        }

        case 'clear-all':
            benchmarkResultAdapter.removeAll(state.totalResults);
            for (const name in state.mount) benchmarkResultAdapter.removeAll(state.mount[name as ComponentName]);
            for (const name in state.unmount) benchmarkResultAdapter.removeAll(state.unmount[name as ComponentName]);
            for (const name in state.update) benchmarkResultAdapter.removeAll(state.update[name as ComponentName]);
            break;

        default:
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            never = action;
    }

    calculateAverageStats(state);
};
