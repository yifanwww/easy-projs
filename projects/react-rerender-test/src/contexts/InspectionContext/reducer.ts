import produce from 'immer';
import type { ImmerReducer } from 'use-immer';

import type { InspectionData } from 'src/common/inspection';

import type { InspectionContextState } from './types';

export type ReducerAction =
    | { type: 'add-record'; groupIndex: number; record: InspectionData }
    | { type: 'register-group'; index: number; group: string }
    | { type: 'toggle-group'; toggle: 'prev' | 'next' };

export const reduce = produce<ImmerReducer<InspectionContextState, ReducerAction>>((draft, action) => {
    let never: never;
    switch (action.type) {
        case 'add-record': {
            const { groupIndex, record } = action;
            const { data, groups } = draft;

            data[groups[groupIndex]].records.push(record);

            break;
        }

        case 'register-group': {
            const { group, index } = action;
            const { data, groups } = draft;

            if (data[group] === undefined) {
                groups.push(group);
                data[group] = {
                    records: [],
                    tree: { group, index, children: [] },
                };
            }

            if (groups.length > 0 && draft.selectedGroup === null) {
                draft.selectedGroup = groups[0];
            }

            break;
        }

        case 'toggle-group': {
            const { toggle } = action;
            const { groups } = draft;

            if (!draft.selectedGroup) break;

            const index = groups.indexOf(draft.selectedGroup);
            if (toggle === 'next') {
                draft.selectedGroup = groups[(index + 1) % groups.length];
            } else {
                draft.selectedGroup = groups[(index + groups.length - 1) % groups.length];
            }

            break;
        }

        /* istanbul ignore next */
        default:
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            never = action;
    }
});
