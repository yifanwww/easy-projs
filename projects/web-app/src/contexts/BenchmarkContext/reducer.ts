import type { Draft } from 'immer';
import type { ImmerReducer } from 'use-immer';

import type { BenchmarkResult, BenchmarkTypes, ComponentName } from 'src/types/benchmark';

import { benchmarkResultAdapter, benchmarkResultSelector } from './adapters';
import type { BenchmarkContextState, GroupBenchmarkResults } from './types';

function calculateAverageStats(draft: Draft<BenchmarkContextState>): void {
    function _calculate(draftGroup: GroupBenchmarkResults, name: ComponentName): void {
        const total = benchmarkResultSelector
            .selectAll(draftGroup[name])
            .reduce((prev, curr) => prev + curr.stats.mean, 0);
        draftGroup.average[name] = total / benchmarkResultSelector.selectTotal(draftGroup[name]);
    }

    _calculate(draft.mount, 'noHooks');
    _calculate(draft.mount, 'useCallback');
    _calculate(draft.mount, 'useMemo');
    _calculate(draft.mount, 'useReducer');
    _calculate(draft.mount, 'useRef');
    _calculate(draft.mount, 'useState');

    _calculate(draft.unmount, 'noHooks');
    _calculate(draft.unmount, 'useCallback');
    _calculate(draft.unmount, 'useMemo');
    _calculate(draft.unmount, 'useReducer');
    _calculate(draft.unmount, 'useRef');
    _calculate(draft.unmount, 'useState');

    _calculate(draft.update, 'noHooks');
    _calculate(draft.update, 'useCallback');
    _calculate(draft.update, 'useMemo');
    _calculate(draft.update, 'useReducer');
    _calculate(draft.update, 'useRef');
    _calculate(draft.update, 'useState');
}

type ReducerAction =
    | { type: 'add'; result: BenchmarkResult }
    | { type: 'clear'; benchmarkType: BenchmarkTypes; componentName: ComponentName }
    | { type: 'clear-all' };

export const reducer: ImmerReducer<BenchmarkContextState, ReducerAction> = (draft, action) => {
    let never: never;
    switch (action.type) {
        case 'add': {
            const { result } = action;

            benchmarkResultAdapter.addOne(draft.totalResults, result);
            benchmarkResultAdapter.addOne(draft[result.type][result.name], result);

            break;
        }

        case 'clear': {
            const { benchmarkType, componentName } = action;

            const selectIds = benchmarkResultSelector.selectIds(draft.totalResults).filter((id) => {
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                const result = benchmarkResultSelector.selectById(draft.totalResults, id)!;
                return result.name === componentName && result.type === benchmarkType;
            });

            benchmarkResultAdapter.removeMany(draft.totalResults, selectIds);
            benchmarkResultAdapter.removeAll(draft[benchmarkType][componentName]);

            break;
        }

        case 'clear-all':
            benchmarkResultAdapter.removeAll(draft.totalResults);
            for (const name in draft.mount) benchmarkResultAdapter.removeAll(draft.mount[name as ComponentName]);
            for (const name in draft.unmount) benchmarkResultAdapter.removeAll(draft.unmount[name as ComponentName]);
            for (const name in draft.update) benchmarkResultAdapter.removeAll(draft.update[name as ComponentName]);
            break;

        /* istanbul ignore next */
        default:
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            never = action;
    }

    calculateAverageStats(draft);
};
