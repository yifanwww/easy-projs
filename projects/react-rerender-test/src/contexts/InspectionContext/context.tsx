import { useForceUpdate } from '@easy-pkg/hooks';
import { abstractFn } from '@easy-pkg/utils';
import { createContext, useCallback, useMemo, useRef } from 'react';

import { useDoubleRenderSign } from 'src/hooks/useDoubleRenderSign';

import { reduce } from './reducer';
import type { ReducerAction } from './reducer';
import type { InspectionContextState, InspectionContextUpdaters } from './types';

export const initialState: InspectionContextState = {
    data: {},
    groups: [],
    selectedGroup: null,
};

export const InspectionContext = createContext<InspectionContextState>(initialState);

export const InspectionContextUpdater = createContext<InspectionContextUpdaters>({
    addRecord: abstractFn,
    forceUpdate: abstractFn,
    registerGroup: abstractFn,
    toggleGroup: abstractFn,
});

export const InspectionProvider: React.FC = ({ children }) => {
    const ref = useRef(initialState);

    const { sign } = useDoubleRenderSign();
    const forceUpdate = useForceUpdate();

    const dispatch = useCallback<React.Dispatch<ReducerAction>>((action) => {
        ref.current = reduce(ref.current, action);
    }, []);

    const updaters = useMemo<InspectionContextUpdaters>(
        () => ({
            addRecord: (record, groupIndex) => {
                if (sign(record)) {
                    dispatch({ type: 'add-record', groupIndex, record });
                }
            },
            forceUpdate,
            registerGroup: (group, index) => dispatch({ type: 'register-group', group, index }),
            toggleGroup: (toggle) => {
                dispatch({ type: 'toggle-group', toggle });
                forceUpdate();
            },
        }),
        [dispatch, forceUpdate, sign],
    );

    return (
        <InspectionContextUpdater.Provider value={updaters}>
            <InspectionContext.Provider value={ref.current}>{children}</InspectionContext.Provider>
        </InspectionContextUpdater.Provider>
    );
};
