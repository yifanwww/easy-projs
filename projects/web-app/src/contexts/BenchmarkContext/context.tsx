import { abstractFn } from '@easy-pkg/utils';
import type { ReactChildrenProps } from '@easy-pkg/utils-react';
import { createContext, useMemo } from 'react';
import { useImmerReducer } from 'use-immer';

import { benchmarkResultAdapter } from './adapters';
import { reducer } from './reducer';
import type { BenchmarkContextState, BenchmarkContextUpdaters } from './types';

export const initialState: BenchmarkContextState = {
    mount: {
        average: { noHooks: 0, useCallback: 0, useMemo: 0, useReducer: 0, useRef: 0, useState: 0 },
        noHooks: benchmarkResultAdapter.getInitialState(),
        useCallback: benchmarkResultAdapter.getInitialState(),
        useMemo: benchmarkResultAdapter.getInitialState(),
        useReducer: benchmarkResultAdapter.getInitialState(),
        useRef: benchmarkResultAdapter.getInitialState(),
        useState: benchmarkResultAdapter.getInitialState(),
    },
    unmount: {
        average: { noHooks: 0, useCallback: 0, useMemo: 0, useReducer: 0, useRef: 0, useState: 0 },
        noHooks: benchmarkResultAdapter.getInitialState(),
        useCallback: benchmarkResultAdapter.getInitialState(),
        useMemo: benchmarkResultAdapter.getInitialState(),
        useReducer: benchmarkResultAdapter.getInitialState(),
        useRef: benchmarkResultAdapter.getInitialState(),
        useState: benchmarkResultAdapter.getInitialState(),
    },
    update: {
        average: { noHooks: 0, useCallback: 0, useMemo: 0, useReducer: 0, useRef: 0, useState: 0 },
        noHooks: benchmarkResultAdapter.getInitialState(),
        useCallback: benchmarkResultAdapter.getInitialState(),
        useMemo: benchmarkResultAdapter.getInitialState(),
        useReducer: benchmarkResultAdapter.getInitialState(),
        useRef: benchmarkResultAdapter.getInitialState(),
        useState: benchmarkResultAdapter.getInitialState(),
    },
    totalResults: benchmarkResultAdapter.getInitialState(),
};

export const BenchmarkContext = createContext<BenchmarkContextState>(initialState);

export const BenchmarkContextUpdater = createContext<BenchmarkContextUpdaters>({
    add: abstractFn,
    clear: abstractFn,
    clearAll: abstractFn,
});

export function BenchmarkProvider({ children }: ReactChildrenProps): JSX.Element {
    const [context, dispatch] = useImmerReducer(reducer, initialState);

    const updaters = useMemo<BenchmarkContextUpdaters>(
        () => ({
            add: (result) => dispatch({ type: 'add', result }),
            clear: (benchmarkType, componentName) => dispatch({ type: 'clear', benchmarkType, componentName }),
            clearAll: () => dispatch({ type: 'clear-all' }),
        }),
        [dispatch],
    );

    return (
        <BenchmarkContextUpdater.Provider value={updaters}>
            <BenchmarkContext.Provider value={context}>{children}</BenchmarkContext.Provider>
        </BenchmarkContextUpdater.Provider>
    );
}
