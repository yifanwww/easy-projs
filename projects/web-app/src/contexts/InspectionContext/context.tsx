import { useForceUpdate } from '@easy-pkg/hooks';
import { abstractFn } from '@easy-pkg/utils';
import type { ReactChildrenProps } from '@easy-pkg/utils-react';
import { createContext, useCallback, useMemo, useRef } from 'react';

import { reduce } from './reducer';
import type { ReducerAction } from './reducer';
import type { InspectionContextState, InspectionContextUpdaters } from './types';

export const initialState: InspectionContextState = {
    data: {},
    groups: [],
    selectedGroup: undefined,
};

export const InspectionContext = createContext<InspectionContextState>(initialState);

export const InspectionContextUpdater = createContext<InspectionContextUpdaters>({
    addRecord: abstractFn,
    forceUpdate: abstractFn,
    registerGroup: abstractFn,
    toggleGroup: abstractFn,
});

export function InspectionProvider({ children }: ReactChildrenProps): JSX.Element {
    const ref = useRef(initialState);

    const forceUpdate = useForceUpdate();

    const dispatch = useCallback<React.Dispatch<ReducerAction>>((action) => {
        ref.current = reduce(ref.current, action);
    }, []);

    const updaters = useMemo<InspectionContextUpdaters>(
        () => ({
            addRecord: (record, groupIndex) => {
                dispatch({ type: 'add-record', groupIndex, record });
            },
            forceUpdate,
            registerGroup: (group, index) => dispatch({ type: 'register-group', group, index }),
            toggleGroup: (toggle) => {
                dispatch({ type: 'toggle-group', toggle });
                forceUpdate();
            },
        }),
        [dispatch, forceUpdate],
    );

    return (
        <InspectionContextUpdater.Provider value={updaters}>
            <InspectionContext.Provider value={ref.current}>{children}</InspectionContext.Provider>
        </InspectionContextUpdater.Provider>
    );
}
