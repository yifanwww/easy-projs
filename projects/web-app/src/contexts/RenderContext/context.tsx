import type { ReactChildrenProps } from '@easy-pkg/helpers-react';
import { abstractFn } from '@easy-pkg/utils';
import { createContext, useMemo } from 'react';
import { useImmerReducer } from 'use-immer';

import { reducer } from './reducer';
import type { RenderContextState, RenderContextUpdaters, RenderContextValues } from './types';

export function getInitialState(): RenderContextState {
    return {
        forceUpdateNumber: 0,
        selected: 0,
    };
}

function getInitialValues(): RenderContextValues {
    return {
        ...getInitialState(),
        forceUpdate: abstractFn,
        select: abstractFn,
    };
}

export const RenderContext = createContext<RenderContextValues>(getInitialValues());

export function RenderProvider({ children }: ReactChildrenProps): React.ReactNode {
    const [state, dispatch] = useImmerReducer(reducer, getInitialState());

    const updaters = useMemo<RenderContextUpdaters>(
        () => ({
            forceUpdate: () => dispatch({ type: 'forceUpdate' }),
            select: (select) => dispatch({ type: 'select', select }),
        }),
        [dispatch],
    );

    return (
        <RenderContext.Provider value={useMemo(() => ({ ...state, ...updaters }), [state, updaters])}>
            {children}
        </RenderContext.Provider>
    );
}
