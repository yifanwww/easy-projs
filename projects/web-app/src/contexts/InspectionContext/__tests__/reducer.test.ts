import type { InspectionData } from 'src/types/inspection';

import { getInitialState } from '../context';
import { reduce } from '../reducer';

describe(`Test reducer fn \`${reduce.name}\``, () => {
    const groups: { index: number; group: string }[] = [
        { index: 0, group: 'group0' },
        { index: 1, group: 'group1' },
        { index: 2, group: 'group2' },
        { index: 3, group: 'group3' },
        { index: 4, group: 'group4' },
    ];

    const records: { groupIndex: number; record: InspectionData }[] = [
        { groupIndex: 0, record: { index: 0, key: 'record0', name: 'record0', parents: [] } },
        { groupIndex: 1, record: { index: 1, key: 'record1', name: 'record1', parents: [] } },
        { groupIndex: 2, record: { index: 2, key: 'record2', name: 'record2', parents: [] } },
        { groupIndex: 2, record: { index: 3, key: 'record3', name: 'record3', parents: [] } },
    ];

    it('should register new group', () => {
        let state = getInitialState();
        for (const group of groups) {
            state = reduce(state, { type: 'register-group', index: group.index, group: group.group });
        }
        expect(state).toMatchSnapshot();
    });

    it('should add new record', () => {
        let state = getInitialState();

        for (const group of groups) {
            state = reduce(state, { type: 'register-group', index: group.index, group: group.group });
        }

        for (const record of records) {
            state = reduce(state, { type: 'add-record', groupIndex: record.groupIndex, record: record.record });
        }

        expect(state).toMatchSnapshot();
    });

    it('should toggle a group', () => {
        // TODO
    });
});
