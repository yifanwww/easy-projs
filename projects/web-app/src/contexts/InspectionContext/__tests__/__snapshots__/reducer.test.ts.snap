// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test reducer fn \`curriedProduce\` should add new record 1`] = `
{
  "data": {
    "group0": {
      "records": [
        {
          "index": 0,
          "key": "record0",
          "name": "record0",
          "parents": [],
        },
      ],
      "tree": {
        "children": [],
        "group": "group0",
        "index": 0,
      },
    },
    "group1": {
      "records": [
        {
          "index": 1,
          "key": "record1",
          "name": "record1",
          "parents": [],
        },
      ],
      "tree": {
        "children": [],
        "group": "group1",
        "index": 1,
      },
    },
    "group2": {
      "records": [
        {
          "index": 2,
          "key": "record2",
          "name": "record2",
          "parents": [],
        },
        {
          "index": 3,
          "key": "record3",
          "name": "record3",
          "parents": [],
        },
      ],
      "tree": {
        "children": [],
        "group": "group2",
        "index": 2,
      },
    },
    "group3": {
      "records": [],
      "tree": {
        "children": [],
        "group": "group3",
        "index": 3,
      },
    },
    "group4": {
      "records": [],
      "tree": {
        "children": [],
        "group": "group4",
        "index": 4,
      },
    },
  },
  "groups": [
    "group0",
    "group1",
    "group2",
    "group3",
    "group4",
  ],
  "selectedGroup": "group0",
}
`;

exports[`Test reducer fn \`curriedProduce\` should register new group 1`] = `
{
  "data": {
    "group0": {
      "records": [],
      "tree": {
        "children": [],
        "group": "group0",
        "index": 0,
      },
    },
    "group1": {
      "records": [],
      "tree": {
        "children": [],
        "group": "group1",
        "index": 1,
      },
    },
    "group2": {
      "records": [],
      "tree": {
        "children": [],
        "group": "group2",
        "index": 2,
      },
    },
    "group3": {
      "records": [],
      "tree": {
        "children": [],
        "group": "group3",
        "index": 3,
      },
    },
    "group4": {
      "records": [],
      "tree": {
        "children": [],
        "group": "group4",
        "index": 4,
      },
    },
  },
  "groups": [
    "group0",
    "group1",
    "group2",
    "group3",
    "group4",
  ],
  "selectedGroup": "group0",
}
`;
